### PR TITLE
EVG-12752: Create REST routes to access system metrics

### PR DESCRIPTION
**Pull request by julianedwards**
JIRA: https://jira.mongodb.org/browse/EVG-12752

Changes: 
- Change the rest/data system metrics mock connector to not use the download method (which requires setting up an env).
- Add route to fetch system metrics by metric type.

### CODE REVIEW

File: rest/system_metrics_routes.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	if len(vals[execution]) > 0 {
+		h.opts.Execution, err = strconv.Atoi(vals[execution][0])
+		return err
+	} else {
+		h.opts.EmptyExecution = true
+	}
+	return nil
+}
+
+// Run finds and returns the desired test result based on task id.
```
Comment (Kimchelly) [inline rest/system_metrics_routes.go:55]: based on task ID and metric type

File: rest/system_metrics_routes.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// Factory returns a pointer to a new systemMetricsGetByTypeHandler.
+func (h *systemMetricsGetByTypeHandler) Factory() gimlet.RouteHandler {
+	return &systemMetricsGetByTypeHandler{
+		sc: h.sc,
+	}
+}
+
+// Parse fetches the task_id metric type from the http request.
```
Comment (Kimchelly) [inline rest/system_metrics_routes.go:39]: task id and metric type

File: rest/system_metrics_routes.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+func (h *systemMetricsGetByTypeHandler) Parse(_ context.Context, r *http.Request) error {
+	var err error
+
+	h.opts.TaskID = gimlet.GetVars(r)["task_id"]
+	h.metricType = gimlet.GetVars(r)["type"]
+	vals := r.URL.Query()
+	if len(vals[execution]) > 0 {
+		h.opts.Execution, err = strconv.Atoi(vals[execution][0])
+		return err
```
Comment (Kimchelly) [inline rest/system_metrics_routes.go:48]: Should this only return if `err != nil`?

File: rest/system_metrics_routes.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+func makeGetSystemMetricsByType(sc data.Connector) gimlet.RouteHandler {
+	return &systemMetricsGetByTypeHandler{
+		sc: sc,
+	}
+}
+
+// Factory returns a pointer to a new systemMetricsGetByTypeHandler.
+func (h *systemMetricsGetByTypeHandler) Factory() gimlet.RouteHandler {
+	return &systemMetricsGetByTypeHandler{
```
Comment (Kimchelly) [inline rest/system_metrics_routes.go:34]: Out of curiosity, it seems like a lot of the REST code does this, but why not just return `makeGetSystemMetricsByType(h.sc)`?

File: rest/system_metrics_routes.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		grip.Error(message.WrapError(err, message.Fields{
+			"request":     gimlet.GetRequestID(ctx),
+			"method":      "GET",
+			"route":       "/system_metrics/type/{task_id}/{type}",
+			"task_id":     h.opts.TaskID,
+			"metric_type": h.metricType,
+		}))
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+	return gimlet.NewJSONResponse(systemMetrics)
```
Comment (Kimchelly) [inline rest/system_metrics_routes.go:69]: `FindSystemMetricsByType` seems to return an `io.ReadCloser` so it seems like you have to read out the results from it.

File: rest/system_metrics_routes_test.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	// no execution
+	rh := s.rh["type"]
+	rh.(*systemMetricsGetByTypeHandler).opts.TaskID = "task1"
+	rh.(*systemMetricsGetByTypeHandler).opts.EmptyExecution = true
+	rh.(*systemMetricsGetByTypeHandler).metricType = "uptime"
+
+	resp := rh.Run(context.TODO())
+	s.Require().NotNil(resp)
+	s.Equal(http.StatusOK, resp.Status())
+	r, ok := resp.Data().(io.ReadCloser)
```
Comment (Kimchelly) [inline rest/system_metrics_routes_test.go:135]: I find this test to be a bit odd - if you're serializing the payload as JSON and sending it over the network, won't the end user receive nothing?

File: rest/system_metrics_routes.go (hunk 4 above)
Comment (julianedwards) [reply rest/system_metrics_routes.go:34]: Not actually sure, something I always copied and pasted lol

File: rest/data/system_metrics.go [hunk 7]
```diff
@@ -48,14 +50,18 @@ func (dbc *DBConnector) FindSystemMetricsByType(ctx context.Context, metricType
 		}
 	}
 
-	return r, nil
+	catcher := grip.NewBasicCatcher()
+	data, err := ioutil.ReadAll(r)
```
Comment (Kimchelly) [inline rest/data/system_metrics.go:54]: Make a JIRA ticket to fix this.

Comment (julianedwards) [reply rest/data/system_metrics.go:54]: Done

**Pull request merged**